### DISCUSSION TITLE
Text shaping & Sitelen Pona

### DISCUSSION THREAD

**Discussion by sowelipililimute** [harfbuzz/harfbuzz#4634] [Q&A] [answered]
## (background context)

I and many others have been working on a [proposal for an ideographic script called Sitelen Pona in Unicode](https://docs.google.com/document/d/1FEVc8fFUEwaSBLBnLXUgyYXKi2wFB04S49mjS98Wk8A/edit). Part of this work has been defining an encoding for two of Sitelen Pona’s most distinguishing features from other logographies: extended glyphs and cartouches. Extended glyphs are glyphs that are extended to “reach below” other glyphs, usually helping to disambiguate grammatical groupings of words.

The following is an example of long pi, the most common word to write extended in Sitelen Pona. (It is the L-shaped glyph going below the others.)

![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/f6a209a5-1bc8-4c82-b029-5dd7d78aeb56)
jan pi __toki pona__ li jan pi __pona mute__
(“a __toki pona__ person is a __very good__ person”)

Cartouches surround names transcribed in Sitelen Pona.
![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/19bc14c2-05f7-4132-8172-5452e2fb3808)
ma [ **k**asi **a**lasa **n**asin **a**wen **t**elo **a** ] li suli

Cartouches and other extended glyphs can be nested within extended glyphs. (The proposed encoding technically allows for nesting cartouches, but this isn't attested in usage.)

![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/ab47b623-3a49-423a-bea2-49c942f44086)
jan pi __ma [ **o** **s**uli **u**ta ]__
(“person of __the land of **Osu**__”)

![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/f1bcd2ad-4fef-48bb-88fc-0e16de93ae99)
jan pi __ma pona pi **toki pona**__
(“person of __the good place of **toki pona**__”)

This is pretty straightforward from an encoding perspective, effectively being nothing more than nestable start/end markers.

## (the problem)

Problems arise from actually implementing these features in fonts. Most current fonts use a simple approach: hardcoded permutations for different scenarios. For example, the glyph “pona” may have the following variations in a font: base, base with pi underline, base with cartouche, base with reverse ala underline, base with forward ala underline, etc. The main technical issue with permutations is a lack of flexibility to accommodate all of the attested usages of Sitelen Pona, as writers combine extended glyphs and cartouches in uncountably many ways, and even a finite subset can push up against OpenType’s glyph limit, if not completely exceed it. 

## (the requirements of a solution)
From what I understand, what’s necessary for “100% accurate” Sitelen Pona is:
- the ability to reposition glyphs with respect to other glyphs (we have multiple GPOS tables to potentially use for this, but I understand that application compatibility regarding whether or not to apply these is iffy)

   This would be used to build extended glyphs/cartouches by e.g. having a “pi.line” glyph that glyphs are positioned on top of, and that can be stacked to build glyphs as deeply nested as desired:

    ![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/b596e72b-c55e-44c4-8b81-4b3edb9c6876)

- something to deal with variable heights that depend on what exactly is enclosed by an extended glyph or cartouche
   ![grafik](https://github.com/harfbuzz/harfbuzz/assets/20326855/21b1d332-60fc-48d4-ba42-44b7769e6494)

My assessment is that whilst implementing positioning of glyphs within nested extensions/cartouches is doable with current OpenType capabilities, implementing variable-height glyphs that are sized with respect to adjacent glyphs is not currently possible and would need something like Beyond Emulation’s `Wasm` table or Graphite in order to implement correctly.

## (the inquiry)

So, my inquiry is less of a question (because I'm not even sure what questions to ask) and more of a request for a sanity check from people that know this stuff way better than any of us Sitelen Pona users. Does this assessment of what is and isn’t possible and how make sense, or am I missing something/reading something incorrectly?

**Comment by plokmijnuhby** [comment]
For the first problem, it seems like GPOS tables are supported, if you use the right feature. Certainly harfbuzz will always support them, and other applications seem to as well.
Variable height glyphs are trickier. One option is to split the cartouche edges into seperate glyphs for the top, bottom, and middle of the edge, and repeat the middle section as needed - you'd also need complicated GSUB and GPOS setups to glue everything together.

**Reply by behdad** [reply]
Right. We also have the `stch` feature support, currently only in the Arabic shaper, which we might be able to reuse here.

**Reply by behdad** [reply]
cc @tiroj

**Reply by plokmijnuhby** [reply]
That's good for solving the first problem, at least. It seems like it doesn't support stretching perpendicular to the text though, so no help at all on the second problem.

**Reply by plokmijnuhby** [reply]
I think the second problem may be beyond the limits of what font software can achieve, if you want to support infinite nesting, but will be possible if you set a finite limit. To determine how many glyphs are needed to draw the initial pi glyph, you'd most likely need to run through the input keeping a running count of how nested you are, either using glyphs with successively higher values, or successive recursions of a contextual lookup. The first option is limited by the maximum number of glyphs in the OpenType spec (which is 2^16), and the second isn't technically limited by the spec but is limited by harfbuzz itself.

**Reply by behdad** [reply]
cc @simoncozens

**Reply by plokmijnuhby** [reply]
Oh, also I read the instructions about `stch` more carefully, and it seems that won't be helpful after all - it does not move glyphs vertically, it merely draws glyphs on top of each other. So we can disregard that.

**Reply by simoncozens** [reply]
By the time this is in Unicode, the WASM shaper will be available. ;-)

**Reply by plokmijnuhby** [reply]
Actually, it already is in unicode, sort of - in the toki pona community there's a spec for how the Private Use Area can be used for sitelen pona. It seems like the WASM shaper is the most straightforward solution here, it's true.

**Reply by tiroj** [reply, marked as answer]
I can imagine a few ways in which this could be implemented with GSUB and GPOS. The important thing for any cartouche or nesting behaviour is having clearly defined and unambiguous start and end delimiters. Ideally, these come from the plain text encoding itself, rather than needing to be contextually arrived at in the glyph processing.

So let’s say you have some characters like [ and ] that indicate the beginning and end of a cartouche, and a character like L that is the Sitelen Pona long pi char but that also indicates the beginning of an extended form that underlines one or more subsequent characters (including potentially a cartouche). You need some way to terminate the long pi extension.

The cartouch and pi line extensions could be handled using width variants, but it makes more sense to me, in terms of flexibility, to break them down into horizontal segments, as suggested by this illustration:
<img width="838" alt="image" src="https://github.com/harfbuzz/harfbuzz/assets/5614500/c97e51d3-9c03-4cdc-b7c7-6b1499696eaa">

The trick in any complex OpenType Layout involving vertical stacking of elements is to determine which of the elements carries the advance width, i.e. which glyph has non-zero spacing, and which are then positione relative to that width. From my understanding of the illustrations, I am going to suggest that the spacing element might change depending on what kind of layout is involved: there are situations in which the spacing element will be the letter, situations in which it is the extending pi line segment, and situations in which it is the cartouche segment. This means you probably need at least two variants of most glyphs: one spacing that acts as a base, and one non-spacing that is classed as a mark and positioned using anchors. For multiple levels of nesting, you may want multiple mark variants of glyphs, as this will enable you to contextually substitute them within nested delimiters. So, for example, a [ begin-cartouche trigger anywhere following the L begin-long-pi trigger but before the close-long-pi trigger would begin the substitution sequence for a set of variants specific to that level of nesting, ending with the next ] close-cartouche trigger. In order to process these substitutions, OTL may require the triggers themselves to be classed as marks so that they can be filtered and processed independently of spacing bases, allowing the contextual substutions to work outwards from the inside of the nest.

Fun!

**Reply by behdad** [reply]
You don't necessarily need two version of each glyph (one mark, one nonmark). You can simply add back the advance width to whichever glyph should carry it.

**Reply by tiroj** [reply]
Or you could always have an invisible spacing background glyph and treat everything else as an anchored mark, but I suspect you _might_ still want separate glyph sets for nesting levels, e.g. smaller forms of letters to fit within nested situations within cartouches, shorter vertical height cartouche segments for nested cartouches, etc.. I have a feeling that for this kind of layout I might just want the convenience of working with separate glyph sets within each nesting level, rather than juggling GPOS everywhere.

**Reply by plokmijnuhby** [reply]
> Ideally, these come from the plain text encoding itself, rather than needing to be contextually arrived at in the glyph processing.

That's one idea, although that wouldn't be consistent with how sitelen pona is currently used. The height of each glyph could be determined by something in the encoding.

**Reply by sowelipililimute** [reply]
the current UCSUR encoding for extended glyphs is `glyph extend-start … extend-end`, and the encoding for cartouches is `cartouche-start … cartouche-end` and this is the mechanism currently being proposed to unicode ATM

**Reply by tiroj** [reply]
> The height of each glyph could be determined by something in the encoding.

That’s probably not needed, or useful. Determining the relative size and position of things is where glyph level processing is appropriate, and trying to manage aspects of that in the encoding—by which I mean character-level—may introduce restrictions that later get in the way of flexible solutions and specific typeface designs, as well as being unlikely to get past Unicode’s technical reviewers.

Where a cartouche begins and ends is semantic content, and I think something similar can be said of the extension of the long pi. So it makes sense for that to be captured in the plain text of the character encoding, and is then useful to glyph processing by providing delimiters to some specific kinds of substition and positioning operations in the font.
